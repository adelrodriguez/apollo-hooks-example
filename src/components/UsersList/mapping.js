export function mapUsers(data) {
  const reducer = (accumulator, value) => {
    if (!value.done) {
      return accumulator + 1;
    }

    return accumulator;
  };

  return data.allUsers.map(user => ({
    id: user.id,
    name: user.name,
    pending: user.Todos.reduce(reducer, 0),
  }));
}

export default {
  mapUsers,
};
