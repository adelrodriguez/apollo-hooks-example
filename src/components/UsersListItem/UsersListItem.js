import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/styles';
import {
  ListItem,
  ListItemText,
  ListItemAvatar,
  ListItemSecondaryAction,
  IconButton,
  Collapse,
  Avatar,
} from '@material-ui/core/';
import DeleteIcon from '@material-ui/icons/Delete';
import Alert from '../Alert';
import TaskList from '../TaskList';

const useStyles = makeStyles({
  avatar: {
    margin: 10,
  },
});

export default function UsersListItem({ user, onRemove }) {
  const classes = useStyles();
  const [collapse, setCollapse] = useState(true);
  const [open, setOpen] = useState(false);

  async function handleRemove() {
    setOpen(false);
    onRemove();
  }

  return (
    <>
      <ListItem button onClick={() => setCollapse(!collapse)}>
        <ListItemAvatar>
          <Avatar className={classes.avatar}>{user.name[0]}</Avatar>
        </ListItemAvatar>
        <ListItemText primary={user.name} secondary={`Pending: ${user.pending}`} />
        <ListItemSecondaryAction>
          <IconButton edge='end' onClick={() => setOpen(true)}>
            <DeleteIcon />
          </IconButton>
        </ListItemSecondaryAction>
      </ListItem>
      <Collapse in={!collapse} timeout='auto' unmountOnExit>
        <TaskList user={user} />
      </Collapse>
      <Alert
        open={open}
        onClose={() => setOpen(false)}
        onAction={handleRemove}
        title='Remove user'
        text={`Are you sure you want to remove the user "${user.name}"?`}
        primaryActionText='Remove'
      />
    </>
  );
}

UsersListItem.propTypes = {
  user: PropTypes.shape({
    id: PropTypes.string,
    name: PropTypes.string,
  }),
  onRemove: PropTypes.func,
};

UsersListItem.defaultProps = {
  user: {
    id: 0,
    name: '',
  },
  onRemove: () => {},
};
