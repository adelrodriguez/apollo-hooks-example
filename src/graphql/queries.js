import { gql } from 'apollo-boost';

export const FETCH_USERS = gql`
  query FetchUsers {
    allUsers {
      id
      name
      Todos {
        done
      }
    }
  }
`;

export const FETCH_USER = gql`
  query FetchUser($id: ID!) {
    User(id: $id) {
      id
      name
      Todos {
        done
      }
    }
  }
`;

export const FETCH_USER_TODOS = gql`
  query FetchUserTodos($id: ID!) {
    allTodos(filter: { user_id: $id }) {
      id
      task
      done
    }
  }
`;

export default {
  // User queries
  FETCH_USERS,
  FETCH_USER,
  // Todos queries
  FETCH_USER_TODOS,
};

