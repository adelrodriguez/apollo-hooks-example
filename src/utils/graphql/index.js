import { mockResponse, mockQueries } from './mockQueries';

export { mockResponse, mockQueries };
export { default as mockClient } from './mockClient';
export { default as mocks } from './mocks';
export { default as withMockedProvider } from './withMockedProvider';
