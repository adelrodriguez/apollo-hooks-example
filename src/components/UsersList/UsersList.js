import React from 'react';
import PropTypes from 'prop-types';
import { makeStyles } from '@material-ui/core/styles';
import { List, ListSubheader } from '@material-ui/core/';
import UsersListItem from '../UsersListItem';

const useStyles = makeStyles(theme => ({
  root: {
    width: '100%',
    maxWidth: 500,
    backgroundColor: theme.palette.background.paper,
  },
}));

export default function UsersList({ users }) {
  const classes = useStyles();

  return (
    <List
      component='div'
      subheader={<ListSubheader component='div'>Users</ListSubheader>}
      className={classes.root}
    >
      {users.map(user => (
        <UsersListItem user={user} key={user.id} />
      ))}
    </List>
  );
}

UsersList.propTypes = {
  users: PropTypes.array,
};

UsersList.defaultProps = {
  users: [],
};
