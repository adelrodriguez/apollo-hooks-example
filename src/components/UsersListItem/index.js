import React from 'react';
import { useMutation } from '@apollo/react-hooks';
import { FETCH_USERS } from '../../graphql/queries';
import { REMOVE_USER } from '../../graphql/mutations';
import UsersListItem from './UsersListItem';

export default function ({ user }) {
  function update(cache) {
    const { allUsers } = cache.readQuery({ query: FETCH_USERS });

    cache.writeQuery({
      query: FETCH_USERS,
      data: { allUsers: allUsers.filter(u => u.id !== user.id) },
    });
  }

  const [removeUser] = useMutation(REMOVE_USER, {
    variables: { id: user.id },
    update,
  });

  return (
    <UsersListItem
      user={user}
      onRemove={removeUser}
    />
  );
}