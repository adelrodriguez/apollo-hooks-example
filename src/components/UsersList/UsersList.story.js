import React from 'react';
import UsersList from './UsersList';
import withMockedProvider from '../../utils/graphql/withMockedProvider';

const users = [{ id: '0', name: 'Lorem Ipsum', pending: 0 }];

export const UI = () => withMockedProvider(<UsersList users={users} />, []);

export default { title: 'UsersList' };
