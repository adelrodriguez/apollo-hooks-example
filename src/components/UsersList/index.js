import React from 'react';
import { useQuery } from '@apollo/react-hooks';
import { FETCH_USERS } from '../../graphql/queries';
import { mapUsers } from './mapping';
import UsersList from './UsersList';

export default function() {
  const { data, loading, error } = useQuery(FETCH_USERS);

  if (loading || error) {
    return null;
  }

  return <UsersList users={mapUsers(data)} />;
}
