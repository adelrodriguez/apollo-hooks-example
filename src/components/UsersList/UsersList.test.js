import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from '@apollo/react-testing';
import { List } from '@material-ui/core';
import { withMockedProvider, mockQueries } from '../../utils/graphql';
import { updateWrapper } from '../../utils/testing';
import { FETCH_USERS } from '../../graphql/queries'
import UnconnectedUsersList from './UsersList';
import UsersList from '.';
import UsersListItem from '../UsersListItem';

describe('UsersList', () => {
  let mocks;

  beforeAll(async () => {
    mocks = await mockQueries({ FETCH_USERS });
  });

  describe('rendering', () => {
    test('renders unconnected', () => {
      const wrapper = shallow(<UnconnectedUsersList />);

      expect(wrapper).toMatchSnapshot();
    });

    test('renders connected', () => {
      const wrapper = shallow(
        <MockedProvider mocks={[]} addTypename={false}>
          <UsersList />
        </MockedProvider>
      );

      expect(wrapper).toMatchSnapshot();
    })

    test('renders connected', () => {
      const wrapper = shallow(withMockedProvider(<UsersList />));

      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('content', () => {
    test('contains 1 List', () => {
      const wrapper = shallow(<UnconnectedUsersList />);

      expect(wrapper.find(List)).toHaveLength(1);
    });

    test('contains at least 1 UsersListItem', async () => {
      const wrapper = mount(withMockedProvider(<UsersList />, mocks));

      await updateWrapper(wrapper);

      expect(wrapper.find(UsersListItem).length).toBeGreaterThanOrEqual(1);
    });
  });
});