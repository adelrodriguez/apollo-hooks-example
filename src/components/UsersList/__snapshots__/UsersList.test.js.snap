// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UsersList rendering renders connected 1`] = `ShallowWrapper {}`;

exports[`UsersList rendering renders connected 2`] = `ShallowWrapper {}`;

exports[`UsersList rendering renders unconnected 1`] = `ShallowWrapper {}`;
